File: frontend/src/types/ShippingProviderSettings.ts
import { ShippingProvider } from "@/enums/ShippingProvider";

export interface ShippingProviderSettings {
  id: number;
  companyId: number;
  warehouseId: number;
  providerName: ShippingProvider;
  accountId: string;
  apiKey?: string;
  apiSecret?: string;
  apiEndpoint?: string;
  enabled: boolean;
  serviceRegions?: string;
  providerConfig?: Record<string, any>;
  createdBy?: number;
  createdAt?: string;
  updatedBy?: number;
  updatedAt?: string;
}

export interface CreateShippingProviderSettingsRequest {
  companyId: number;
  warehouseId: number;
  providerName: ShippingProvider;
  accountId: string;
  apiKey?: string;
  apiSecret?: string;
  apiEndpoint?: string;
  enabled?: boolean;
  serviceRegions?: string;
  providerConfig?: Record<string, any>;
}

export interface UpdateShippingProviderSettingsRequest {
  providerName?: ShippingProvider;
  accountId?: string;
  apiKey?: string;
  apiSecret?: string;
  apiEndpoint?: string;
  enabled?: boolean;
  serviceRegions?: string;
  providerConfig?: Record<string, any>;
}

File: backend/src/main/java/nextpos/app/nextpos/model/entity/ShippingProviderSettings.java
package nextpos.app.nextpos.model.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;

import java.time.LocalDateTime;
import java.util.Map;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nextpos.app.nextpos.model.enums.ShippingProvider;

import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Entity
@Table(name = "shipping_provider_settings")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
@EntityListeners(AuditingEntityListener.class)
public class ShippingProviderSettings {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @EqualsAndHashCode.Include
    private Long id;

    /** Company still linked (for tenant scoping & access control) */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "company_id", nullable = false)
    private Company company;

    /** Each warehouse/outlet has its own provider setup */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "warehouse_id", nullable = false)
    private Warehouse warehouse;

    @Enumerated(EnumType.STRING)
    @Column(name = "provider_name", nullable = false, length = 50)
    private ShippingProvider providerName;

    @Column(name = "account_id")
    private String accountId;

    @Column(name = "api_key")
    private String apiKey;

    @Column(name = "api_secret")
    private String apiSecret;

    @Column(name = "api_endpoint")
    private String apiEndpoint;

    @Column(name = "enabled", nullable = false)
    @Builder.Default
    private Boolean enabled = Boolean.TRUE;

    /**
     * For which regions this provider works (optional free-text, e.g., "IN,EU,US")
     */
    @Column(name = "service_regions")
    private String serviceRegions;

    /** Store provider-specific settings in JSON */
    @Column(columnDefinition = "jsonb")
    @JdbcTypeCode(SqlTypes.JSON)
    private Map<String, Object> providerConfig;

    @Column(name = "created_by", nullable = false, updatable = false)
    private Long createdBy;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_by")
    private Long updatedBy;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}
